### PR TITLE
On-demand replication Integration Test and more unit tests

### PR DESCRIPTION
**Pull request by JingQianCloud**
1. Added ReplicateBlob Integration Test
2. Added CompositeBlob test for ReplicateBlob request and TtlUpdate retry.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java [hunk 1]
```diff
@@ -248,4 +248,31 @@ public boolean equals(Object o) {
...
+    return isPrivate == that.isPrivate
+        && creationTimeInMs == that.creationTimeInMs
+        && accountId == that.accountId
+        && containerId == that.containerId
+        && isEncrypted == that.isEncrypted
+        && blobSize == that.blobSize
+        && timeToLiveInSeconds == that.timeToLiveInSeconds
+        && Objects.equals(serviceId, that.serviceId)
+        && (Objects.equals(ownerId, that.ownerId) || ((ownerId == null || ownerId.isEmpty()) && (that.ownerId == null || that.ownerId.isEmpty())))
+        && (Objects.equals(contentType, that.contentType) || ((contentType == null || contentType.isEmpty()) && (that.contentType == null || that.contentType.isEmpty())))
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java:274]: should probably create a helper method and also use this helper method to check filename and contentEncoding?

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java [hunk 2]
```diff
@@ -2436,6 +2438,578 @@ static void undeleteRecoveryTest(Port targetPort, MockCluster cluster, SSLConfig
...
+          encryptionKey, clusterMap, blobIdFactory, testEncryption);
+
+      // 2. On the sourceDataNode, putBlob. On the targetDataNode, the blob already exists.
+      // ReplicateBlob to both the sourceDataNode and the targetDataNode. All succeed.
+      blobId = blobId2;
+      putRequest = new PutRequest(1, "client1", blobId, propertiesWithTtl, ByteBuffer.wrap(userMetadata),
+          Unpooled.wrappedBuffer(data), properties.getBlobSize(), BlobType.DataBlob,
+          testEncryption ? ByteBuffer.wrap(encryptionKey) : null);
+      putBlob(putRequest, sourceChannel, ServerErrorCode.No_Error);
+      replicateBlob(sourceChannel, blobId, sourceDataNode, ServerErrorCode.No_Error);
```
Comment (justinlin-linkedin) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2559]: this is a great test. but in AmbryRequests implementation, we should probably check if the source host  is the same host as localhost, if so we should just return true right away.

**Review by justinlin-linkedin — APPROVED** [review]
minor comments

File: ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java (hunk 1 above)
Comment (JingQianCloud) [reply ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java]: Done.

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java (hunk 2 above)
Comment (JingQianCloud) [reply ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2559]: Yes, plan to do that. Will file a separate PR for it.

File: ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java [hunk 3]
```diff
@@ -248,4 +248,41 @@ public boolean equals(Object o) {
         contentEncoding, that.contentEncoding) && Objects.equals(filename, that.filename) && Objects.equals(
         externalAssetTag, that.externalAssetTag);
   }
+
+  /**
+   * If the string is null or empty, return true.
+   * @param str string to check
+   * @return true if the string is either null or is an empty string.
+   */
+  private boolean nullOrEmptry(String str) {
```
Comment (SophieGuo410) [inline ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java:257]: nit: please format this method.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 4]
```diff
@@ -1530,6 +1532,102 @@ public void testReplicateBlobFailureDueToNotEnoughQuorum() throws Exception {
...
+        if (server.getDataCenter().equals(localDcName)) {
+          server.setServerErrorForAllRequests(ServerErrorCode.Replica_Unavailable);
+        }
+      }
+      GetBlobResult remoteGetBlobResult = router.getBlob(compositeBlobId, new GetBlobOptionsBuilder().build(), null, null).get();
+
+      // verify BlobProperties
+      Assert.assertEquals(remoteGetBlobResult.getBlobInfo().getBlobProperties(), localProperties);
+      // verify userMetadata
+      Assert.assertTrue(Arrays.equals(remoteGetBlobResult.getBlobInfo().getUserMetadata(), putUserMetadata));
```
Comment (SophieGuo410) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1610]: Can it be simplify as
`assertArrayEquals(remoteGetBlobResult.getBlobInfo().getUserMetadata(), putUserMetadata)`

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 5]
```diff
@@ -1530,6 +1532,102 @@ public void testReplicateBlobFailureDueToNotEnoughQuorum() throws Exception {
...
+      // verify data content
+      RetainingAsyncWritableChannel retainingAsyncWritableChannel = new RetainingAsyncWritableChannel();
+      remoteGetBlobResult.getBlobDataChannel().readInto(retainingAsyncWritableChannel, null).get();
+      InputStream input = retainingAsyncWritableChannel.consumeContentAsInputStream();
+      retainingAsyncWritableChannel.close();
+      byte[] content = Utils.readBytesFromStream(input, (int) PUT_CONTENT_SIZE);
+      input.close();
+      Assert.assertTrue(Arrays.equals(putContent, content));
```
Comment (SophieGuo410) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1618]: Same here.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 6]
```diff
@@ -1530,6 +1532,102 @@ public void testReplicateBlobFailureDueToNotEnoughQuorum() throws Exception {
...
+      // verify BlobProperties
+      Assert.assertEquals(remoteGetBlobResult.getBlobInfo().getBlobProperties(), localProperties);
+      // verify userMetadata
+      Assert.assertTrue(Arrays.equals(remoteGetBlobResult.getBlobInfo().getUserMetadata(), putUserMetadata));
+      // verify data content
+      RetainingAsyncWritableChannel retainingAsyncWritableChannel = new RetainingAsyncWritableChannel();
+      remoteGetBlobResult.getBlobDataChannel().readInto(retainingAsyncWritableChannel, null).get();
+      InputStream input = retainingAsyncWritableChannel.consumeContentAsInputStream();
+      retainingAsyncWritableChannel.close();
+      byte[] content = Utils.readBytesFromStream(input, (int) PUT_CONTENT_SIZE);
```
Comment (SophieGuo410) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1616]: There's no need to cast since it's int already right?

**Review by SophieGuo410 — APPROVED** [review]
LGTM! Just left some minor comment.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 4 above)
Comment (JingQianCloud) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java]: Good to know that.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 5 above)
Comment (JingQianCloud) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java]: Changed.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 6 above)
Comment (JingQianCloud) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java]: You are right.

File: ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java (hunk 3 above)
Comment (JingQianCloud) [reply ambry-api/src/main/java/com/github/ambry/messageformat/BlobProperties.java:257]: Done.

**Pull request merged**